src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -314,58 +390,92 @@ try_validate_json_schema(const jsoncons::json& schema) {
-            "Malformed json schema: {} at offset {}",
-            rapidjson::GetParseError_En(schema.GetParseError()),
-            schema.GetErrorOffset())};
+            "Malformed json schema: {} at line {} column {}",
+            ec ? ec.message() : "Invalid document",
+            reader.line(),
+            reader.column())};
     }
+    auto schema = decoder.get_result();
 
     // get the dialect, try to directly validate it against the appropriate
     // metaschema
-    auto dialect = std::optional<json_schema_dialect>{};
+    auto maybe_dialect = std::optional<json_schema_dialect>{};
 
-    if (schema.IsObject()) {
+    if (schema.is_object()) {
         // "true/false" are valid schemas so here we need to check that the
         // schema is an actual object
-        if (auto it = schema.FindMember("$schema"); it != schema.MemberEnd()) {
-            if (it->value.IsString()) {
-                dialect = from_uri(as_string_view(it->value));
+        if (auto it = schema.find("$schema");
+            it != schema.object_range().end()) {
+            if (it->value().is_string()) {
+                maybe_dialect = from_uri(it->value().as_string_view());
             }
 
-            if (it->value.IsString() == false || dialect == std::nullopt) {
+            if (
+              it->value().is_string() == false || !maybe_dialect.has_value()) {
                 // if present, "$schema" have to be a string, and it has to be
                 // one the implemented dialects. If not, return an error
                 return error_info{
                   error_code::schema_invalid,
                   fmt::format(
-                    "Unsupported json schema dialect: '{}'", pj{it->value})};
+                    "Unsupported json schema dialect: '{}'",
+                    jsoncons::print(it->value()))};
             }
         }
     }
 
     // We use jsoncons for validating the schema against the metaschema as
     // currently rapidjson doesn't support validating schemas newer than
     // draft 5.
-    iobuf_istream is{std::move(buf)};
-    auto jsoncons_schema = jsoncons::json::parse(is.istream());
-    auto validation_res = dialect.has_value()
+    auto validation_res = maybe_dialect.has_value()
                             ? validate_json_schema(
-                                dialect.value(), jsoncons_schema)
-                            : try_validate_json_schema(jsoncons_schema);
+                                maybe_dialect.value(), schema)
+                            : try_validate_json_schema(schema);
     if (validation_res.has_error()) {
         return validation_res.as_failure();
     }
+    auto dialect = validation_res.assume_value();
+
+    // this function will resolve al local ref against their respective baseuri.
+    auto bundles_schemas_map = collect_bundled_schema_and_fix_refs(
+      schema, dialect);
+
+    // to use rapidjson we need to serialized schema again
+    auto iobuf_os = iobuf_ostream{};
```

Comment:
```suggestion
    // to use rapidjson we need to serialize the schema again
    // We take a copy of the jsoncons schema here because it has the fixed-up references that we want to use for compatibility checks
    auto iobuf_os = iobuf_ostream{};
```

Reply:
done